utils/model_runners/bedrock_counting_runner.py

```diff
@@ -0,0 +1,92 @@
+        """
+        :param model_id: Id of the Bedrock model to be used for model predictions
+        :param content_template: String template to compose the model input from the prompt
+        :param output: JMESPath expression of output in the model output
+        :param log_probability: JMESPath expression of log probability in the model output
+        :param content_type: The content type of the request sent to the model for inference
+        :param accept_type: The accept type of the request sent to the model for inference
+        :param metrics_folder: The destination of the invocation metric file
+        :param model_key: The base name of the file to disambiguate metrics
+        """
+        super().__init__(model_id = model_id, content_template = content_template, output = output, log_probability=log_probability, content_type = content_type, accept_type = accept_type)
+        self._metrics_folder = metrics_folder
+        self._model_key = model_key
+        
+
+
+    def predict(self, prompt: str) -> Tuple[Optional[str], Optional[float]]:
+        """
+        Invoke the Bedrock model and parse the model response.
+        :param prompt: Input data for which you want the model to provide inference.
+        """
+        
+        composed_data = self._composer.compose(prompt)
+        body = json.dumps(composed_data)
+        start_time = datetime.now(timezone.utc)
+
+        response = self._bedrock_runtime_client.invoke_model(
+            body=body, modelId=self._model_id, accept=self._accept_type, contentType=self._content_type
+        )
+        delta =  datetime.now(timezone.utc) - start_time
+        processing_time = delta.total_seconds()
+        model_output = json.loads(response.get("body").read())
+        
+        input_token_count = int(response["ResponseMetadata"]["HTTPHeaders"][
+            "x-amzn-bedrock-input-token-count"
+        ])
+        
+        output_token_count = int(response["ResponseMetadata"]["HTTPHeaders"][
+            "x-amzn-bedrock-output-token-count"
+        ])
+        
+        output = (
+            self._extractor.extract_output(data=model_output, num_records=1)
+            if self._extractor.output_jmespath_expression
+            else None
+        )
+        log_probability = (
+            self._extractor.extract_log_probability(data=model_output, num_records=1)
+            if self._extractor.log_probability_jmespath_expression
+            else None
+        )
+        
+        sw = json.dumps({"input_tokens":input_token_count,"output_tokens":output_token_count, "processing_time":processing_time,"model_id":self._model_id})
+        fp = open(self._metrics_folder + f"/{self._model_key}_usage.jsonl", 'a')
+        fcntl.flock(fp.fileno(), fcntl.LOCK_EX)
```

Comment:
why do you need seek() and flock()  instead of just appending to the end of the file?

Reply:
because between open and lock another process can append to the file, so I need to be sure I'm at the last position. notebook is running parallelism 1, but this way if user tweak parallelism code is still robust